include/circt/Dialect/FIRRTL/FIRRTLExpressions.td

```diff
@@ -662,3 +662,70 @@ def BitCastOp: FIRRTLOp<"bitcast", [NoSideEffect]> {
 
   let assemblyFormat = "$input attr-dict `:` functional-type($input, $result)";
 }
+
+//===----------------------------------------------------------------------===//
+// RefOperations: Operations on the RefType.
+//
+// The current motivation is that these operations will be used to express
+// cross module references (xmr). The RefOps will be lowered to sv.verbatim
+// to encode the xmr.
+// These operations can be used to express only a remote read. Remote write
+// is still not supported.
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// Constraints on RefOps
+//===----------------------------------------------------------------------===//
+
+class RefTypeConstraint<string ref, string base>
+  : TypesMatchWith<"reference base type should match",
+                   ref, base,
+                   "$_self.cast<RefType>().getType()">;
+
+class RefSingleUserConstraint<string ref>
+  : PredOpTrait<"reference port operand cannot be reused by any other op",
+                CPred<"$" # ref # ".hasOneUse()">>;
```

Comment:
`ref` is always `"ref"` right now. Would it be better to make this not take the parameter? Ditto for other constraints below.

Reply:
True, we can remove the `ref` from the parameters.